Hold fetch semaphore units until response is sent

ballard26 opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR does the following:
- Adds a `ss::deleter` to `session_resources` that can be added to
  via `request_context`. This allows request handlers to free any resources
  related to the response after its been fully sent to the client.
- Adds `response_memory_units` to `op_context` to track memory units per 
  partition in the fetch response.
- Modifies `fill_fetch_responses` to ensures units in 
  `op_context::response_memory_units` accurately reflect data being used in 
  the fetch response.
- A few additional changes in `fill_fetch_responses` to make a more accurate
  `bytes_left` calculation and to avoid unnecessarily discarding data in the fetch response.


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

ballard26 committed:
kafka/server: unnest read_result::memory_units_t as fetch_memory_units

ballard26 committed:
kafka/server: access session_resources via request_context

This commit adds a ss::deleter to session_resources that can be added to
via request_context. This allows request handlers to free any resources
related to the response after its been fully sent to the client.

ballard26 committed:
kafka/server: cache ktp in op_context::response_placeholder

ballard26 committed:
kafka/server: add response_size method to op_context::response_placeholder

ballard26 committed:
kafka/server: track response memory units in op_context

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72229-->
<!--90853e6f-93a2-42ab-a847-bde1b4cadcc0-->
<!--title-->
## Retry command for [Build#72229](https://buildkite.com/redpanda/redpanda/builds/72229)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/follower_fetching_test.py::FollowerFetchingTest.test_basic_follower_fetching@{"read_from_object_store":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--362be371-dbe3-4f5b-9a3b-195c18920f53-->
<!--TEST RESULTS-->
<!--68ad6403-8176-401a-acf8-8eee7b5ef5ba-->
<!--TEST RESULTS-->
<!--fca78e4b-d012-4970-92d8-b28c2a18327b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72229">build#72229</a></summary>


| test_class                      | test_method                                   | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:--------------------------------|:----------------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest        | test_scrubber                                 | {"cloud_storage_type": 1}         | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999f-4d71-a53a-dcfcea619656 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| ArroyoTest                      | test_arroyo_test_suite                        | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee9-4c97-a35f-e643fdcbe3a9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| ArroyoTest                      | test_arroyo_test_suite                        | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999d-4b66-8007-d0f1d4ddb94a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_basic_follower_fetching                  | {"read_from_object_store": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee6-4e82-9548-513cfbc0badf | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_basic_follower_fetching                  | {"read_from_object_store": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999a-46ac-845e-b0504dbf7722 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_basic_follower_fetching                  | {"read_from_object_store": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee7-4cbd-b368-cfb6d14d04dd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_basic_follower_fetching                  | {"read_from_object_store": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999b-48f5-b3f1-68b1e4c8a0af | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_follower_fetching_with_maintenance_mode  | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee9-4c97-a35f-e643fdcbe3a9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_follower_fetching_with_maintenance_mode  | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999d-4b66-8007-d0f1d4ddb94a | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL |
| FollowerFetchingTest            | test_with_leadership_transfers                | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1eea-421e-90dd-113b8f173f35 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| FollowerFetchingTest            | test_with_leadership_transfers                | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-999f-4d71-a53a-dcfcea619656 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| IncrementalFollowerFetchingTest | test_incremental_fetch_from_follower          | {"follower_offline": false}       | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1eeb-49e8-993f-a44554ca8ba6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| IncrementalFollowerFetchingTest | test_incremental_fetch_from_follower          | {"follower_offline": true}        | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee1-4f7f-aade-33eb2c1829f6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| StreamVerifierTest              | test_simple_produce_consume_txn_with_add_node | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468b-1ee1-4f7f-aade-33eb2c1829f6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| StreamVerifierTest              | test_simple_produce_consume_txn_with_add_node | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/72229#0199468c-9993-4117-a650-802cf1765d80 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72232">build#72232</a></summary>


| test_class                 | test_method                                   | test_arguments                                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       |
|:---------------------------|:----------------------------------------------|:--------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArroyoTest                 | test_arroyo_test_suite                        | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b640-4f82-8d7b-dc8741b1d906 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                       |
| ArroyoTest                 | test_arroyo_test_suite                        | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3c2-4fb4-b473-04301b5dfb02 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                       |
| KafkaStreamsPageView       | test_kafka_streams_page_view                  | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b642-4249-ab4b-1f1ab6e65ab2 | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL  |
| KafkaStreamsSessionWindow  | test_kafka_streams                            | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3bb-485a-b717-4b225fd45112 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS  |
| KafkaStreamsSumLambda      | test_kafka_streams                            | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3bc-4e89-aa2a-9608f6337006 | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL  |
| KafkaStreamsTopArticles    | test_kafka_streams                            | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b63a-44c8-a807-5bdedd95ff4e | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL |
| KafkaStreamsTopArticles    | test_kafka_streams                            | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3bd-4714-bdba-3364be43d5e0 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL |
| TestVirtualConnections     | test_no_head_of_line_blocking                 | {"different_clusters": false, "different_connections": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3c0-4442-9467-e1b35fcc2dfd | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL |
| TestVirtualConnections     | test_no_head_of_line_blocking                 | {"different_clusters": false, "different_connections": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3c1-4cc2-852a-ed771ac87cea | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL  |
| TestVirtualConnections     | test_no_head_of_line_blocking                 | {"different_clusters": true, "different_connections": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b640-4f82-8d7b-dc8741b1d906 | FLAKY         | 7/21     | upstream reliability is '100.0'. current run reliability is '33.33333333333333'. drift is 66.66667 and the allowed drift is set to 50. The test should FAIL  |
| TestVirtualConnections     | test_no_head_of_line_blocking                 | {"different_clusters": true, "different_connections": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b641-4b2d-89a1-35ea87fb85be | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL |
| TestVirtualConnections     | test_no_head_of_line_blocking                 | {"different_clusters": true, "different_connections": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3c3-4373-98b1-86a15aed6119 | FLAKY         | 9/21     | upstream reliability is '100.0'. current run reliability is '42.857142857142854'. drift is 57.14286 and the allowed drift is set to 50. The test should FAIL |
| DataMigrationsApiTest      | test_mount_inexistent                         | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b63d-439e-8664-e95a5aaf8656 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS    |
| DatalakeClusterRestoreTest | test_slow_tiered_storage_dlq                  | {"catalog_type": "nessie", "cloud_storage_type": 1}           | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3c4-49ad-909c-4ff0bf406bc8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS    |
| StreamVerifierTest         | test_simple_produce_consume_txn_with_add_node | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b86-b638-49b2-a031-a6e7b2938e13 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                       |
| StreamVerifierTest         | test_simple_produce_consume_txn_with_add_node | null                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/72232#01994b87-a3bb-485a-b717-4b225fd45112 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72642">build#72642</a></summary>


| test_class           | test_method                                       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                         |
|:---------------------|:--------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest | test_client_group_consume_rate_throttle_mechanism | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72642#01996df2-ca7e-49d9-8477-827abcaa8970 | FLAKY         | 16/21    | upstream reliability is '98.57142857142858'. current run reliability is '76.19047619047619'. drift is 22.38095 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| ClusterRateQuotaTest | test_client_group_consume_rate_throttle_mechanism | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72642#01996df5-6167-4d79-8e45-76c9c122c16e | FLAKY         | 14/21    | upstream reliability is '93.87755102040816'. current run reliability is '66.66666666666666'. drift is 27.21088 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| ClusterRateQuotaTest | test_client_group_produce_rate_throttle_mechanism | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72642#01996df2-ca74-481c-9496-c688a3df8ceb | FLAKY         | 15/21    | upstream reliability is '97.14964370546319'. current run reliability is '71.42857142857143'. drift is 25.72107 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism |
</details>
<!--/body-->

ballard26 committed:
kafka/server: refactor fill_fetch_responses

This commit makes the following changes to fill_fetch_responses:
- Ensures units in op_context::response_memory_units accurately reflect
  data being used in the fetch response.
- Changes the `bytes_left` calculation when determining if there is
  enough space for a given partition response. Previously
  `octx.bytes_left` was used. However, this value includes the size of
  the last read result written to the response for the partition. Hence
  that pervious result needs to be subtracted out first to get the
  accurate number of bytes left.
- When there isn't enough bytes left in the response to write the read
  result for a given partition we currently skip writing that read
  result and clear any previously written read result for that partition
  from the response. Now we just skip writing the new read result to the
  response and retain any previously written one instead of clearing it.

ballard26 commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -534,27 +555,31 @@ static void fill_fetch_responses(
-            } else {
-                total_memory_units->adopt(std::move(res.memory_units));
-            }
+        if (record_latency) {
```
@travisdowns Wonder if its worth discussing which(if any) fetch latency metrics are useful again. It almost seems like we want a latency per byte read, or similar. Rather than latency per partition read with no idea on the read size. It makes even less sense in polling fetch where the workers may just sleep for a long period of time waiting for the partition to have new data.

ballard26 replied:
Yeah, nothing for now is the decision we reached.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72232-->
<!--e531e937-abc2-4c50-848b-0527b4f8a285-->
<!--465e8272-e633-432a-8af9-b177943f868d-->
<!--5513948e-e386-4119-9077-3b756317bec7-->
<!--98822dbd-1871-4a3b-ac41-c0f1b2040d78-->
<!--4c37e7e7-4ed6-4a5f-8638-a16699a8b72e-->
<!--96cbdabf-bedb-42db-885a-51bc14d539d1-->
<!--29f9d5ff-bb02-4a31-8682-f615b6a1641c-->
<!--2ee46b3d-c7bb-4cae-962c-f07dabe0ef31-->
<!--c4be423d-5d79-4c8d-82fc-0e3079e2ccaf-->
<!--06bfe94b-202d-4b26-bf86-e7354afd6e0f-->
<!--c26eb0d5-d756-46b5-a9b6-0893f617c5f1-->
<!--95ab84ab-e1cd-4a21-97f4-cbd152569157-->
<!--title-->
## Retry command for [Build#72232](https://buildkite.com/redpanda/redpanda/builds/72232)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/connection_virtualizing_test.py::TestVirtualConnections.test_no_head_of_line_blocking@{"different_clusters":true,"different_connections":true}
tests/rptest/tests/connection_virtualizing_test.py::TestVirtualConnections.test_no_head_of_line_blocking@{"different_clusters":false,"different_connections":true}
tests/rptest/tests/compatibility/kafka_streams_test.py::KafkaStreamsSumLambda.test_kafka_streams
tests/rptest/tests/connection_virtualizing_test.py::TestVirtualConnections.test_no_head_of_line_blocking@{"different_clusters":false,"different_connections":false}
tests/rptest/tests/compatibility/arroyo_test.py::ArroyoTest.test_arroyo_test_suite
tests/rptest/tests/compatibility/kafka_streams_test.py::KafkaStreamsTopArticles.test_kafka_streams
tests/rptest/tests/compatibility/kafka_streams_test.py::KafkaStreamsPageView.test_kafka_streams_page_view
tests/rptest/tests/connection_virtualizing_test.py::TestVirtualConnections.test_no_head_of_line_blocking@{"different_clusters":true,"different_connections":false}
tests/rptest/transactions/stream_verifier_test.py::StreamVerifierTest.test_simple_produce_consume_txn_with_add_node
```
<!--/body-->

BenPope commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -68,6 +68,14 @@ struct op_context {
         intrusive_list_hook _hook;
 
+        size_t response_size() {
+            auto& records_opt = _it->partition_response->records;
+            if (!records_opt) {
+                return 0;
+            }
+            return records_opt->size_bytes();
```
nitpick:
```suggestion
            return _it->partition_response->records
              .transform(&kafka::batch_reader::size_bytes)
              .value_or(0);
```

ballard26 replied:
Oh, nice call. Switched to your suggestion.

BenPope commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -173,6 +185,13 @@ struct op_context {
+    // current units for that ktp in order to fill the response with the newly
+    // read result.
+    chunked_hash_map<model::ktp_with_hash, fetch_memory_units>
+      response_memory_units;
```
Is it worth encapsulating this as a type with operations on it?

ballard26 replied:
Done, let me know if it looks better now.

BenPope commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -554,27 +555,31 @@ static void fill_fetch_responses(
+            // It's possible that data for this partition has already been
+            // written to the response, but is no longer in the partition due to
+            // truncation or otherwise. Hence we move any existing response
+            // untis out of the op_context and into the deleter here.
```
```suggestion
            // units out of the op_context and into the deleter here.
```

BenPope reviewed (commented):
LGTM

ballard26 committed:
kafka/server: add test for fetch response memory units

This test ensures that the units being held for a given fetch response
equal the size(in bytes) of that response.

ballard26 commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -505,17 +505,18 @@ ss::future<> connection_context::handle_auth_v0(const size_t size) {
 
```
A DT test for the safe guards will have to wait a bit longer until the changes for obligatory reads are in. For now I've added a fixture test for ensuring that the number of semaphore units that'll be held in the session resources matches the response size. Let me know if that works for you.

vbotbuildovich commented:
<!--7688ed6a-dd72-42a4-ba2d-61f364281c72-->
<!--7bccfccb-c1b2-4711-ade9-9cd18a3d64a5-->
<!--body-->
Performance change detected in https://buildkite.com/redpanda/redpanda/builds/72657#0199718f-e715-4cd3-a224-cc4ca8e5fe20:
```
Performance changes detected in 14 tests
ssx_bench_rpbench_ss_ss_fmt_1K.join: inst -> -0.05pct
kafka_produce_partition_rpbench_produce_partition_fixture.4_KiB_produced: inst -> +34.16pct
kafka_fetch_rpbench_large_fetch_t.multi_partition_fetch_version_max: allocs -> +5.26pct
kafka_fetch_rpbench_large_fetch_t.multi_partition_fetch_version_with_epoch_validation: allocs -> +5.26pct
kafka_fetch_rpbench_large_fetch_t.multi_partition_fetch_version_with_rack: allocs -> +5.29pct
kafka_fetch_rpbench_large_fetch_t.single_partition_fetch_version_max: allocs -> +4.76pct
kafka_fetch_rpbench_large_fetch_t.single_partition_fetch_version_with_epoch_validation: allocs -> +4.76pct
kafka_fetch_rpbench_large_fetch_t.single_partition_fetch_version_with_rack: allocs -> +4.76pct
kafka_fetch_rpbench_small_fetch_t.multi_partition_fetch_version_max: allocs -> +5.74pct
kafka_fetch_rpbench_small_fetch_t.multi_partition_fetch_version_with_epoch_validation: allocs -> +5.74pct
kafka_fetch_rpbench_small_fetch_t.multi_partition_fetch_version_with_rack: allocs -> +5.74pct
kafka_fetch_rpbench_small_fetch_t.single_partition_fetch_version_max: allocs -> +4.76pct
kafka_fetch_rpbench_small_fetch_t.single_partition_fetch_version_with_epoch_validation: allocs -> +4.76pct
kafka_fetch_rpbench_small_fetch_t.single_partition_fetch_version_with_rack: allocs -> +4.76pct
```
See https://redpandadata.atlassian.net/wiki/x/LQAqLg for docs
<!--/body-->

ballard26 commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -1631,13 +1715,19 @@ fetch_handler::handle(request_context rctx, ss::smp_service_group ssg) {
           }
           octx.response.data.error_code = error_code::none;
           return do_fetch(octx).then([&octx] {
+              auto resp_units_deleter
```
They're technically being moved from the op_context to the request_context/session resources. And they're not aggregated by shard in the first place due to us needing to free/replace units per ktp throughout the fetch. After the response has been finalized though we don't need the ktp info and can aggregate them by shard for cheaper destruction.

pgellert commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -37,6 +38,29 @@ using fetch_handler = single_stage_handler<
+    ss::deleter into_deleter() &&;
+
+    /// Returns the total number of units held by this class.
+    size_t unit_count();
```
nit: could be `const`

ballard26 replied:
Will change

pgellert commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -68,6 +68,12 @@ struct op_context {
         }
         intrusive_list_hook _hook;
 
+        size_t response_size() {
```
nit: could be const

ballard26 replied:
Will make it const

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -235,6 +235,73 @@ void fetch_memory_units::adopt(fetch_memory_units&& o) {
+      [](auto v) { return std::get<1>(std::move(v)); });
+}
+
+ss::deleter memory_units_by_ktp::into_deleter() && {
+    return ss::make_deleter([memory_units = std::move(_memory_units)] mutable {
+        // Aggregate units to ensure only one cross shard call to each shard at
+        // most.
```
I'm new to `ss::deleter` so I have a question here. Is this lambda ever getting called? The [docs](https://docs.seastar.io/master/group__memory-module.html#ga8a3eb323c7e666c1dbe8f05f1eae9ec7) would make me believe that here we are just creating a lambda object that captures what it captures and is going to be freed when the deleter is triggered, so I would expect that the lambda code won't ever get called. I.e. this code would be equivalent to just
```
return ss::make_deleter(std::move(_memory_units));
```

What am I missing?

ballard26 replied:
The documentation really isn't great. AFAIK the following is called:
```c++
  template <typename Object>
  deleter
  make_deleter(Object o) {
      return make_deleter(deleter(), std::move(o));
  }
  ```
Which in turn calls:
```c++
  template <typename Object>
  deleter
  make_deleter(deleter next, Object o) {
      return deleter(new lambda_deleter_impl<Object>(std::move(next), std::move(o)));
  }
```
Which creates a deleter with the following impl;
```c++
  /// \cond internal
  template <typename Deleter>
  struct lambda_deleter_impl final : deleter::impl {
      Deleter del;
      lambda_deleter_impl(deleter next, Deleter&& del)
          : impl(std::move(next)), del(std::move(del)) {}
      virtual ~lambda_deleter_impl() override { del(); }
  };
```
so the lambda should be called. There is also `ss::make_object_deleter` which does that you are thinking and only destroys the passed object on deletion.

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -235,6 +235,73 @@ void fetch_memory_units::adopt(fetch_memory_units&& o) {
+    void add_units(fetch_memory_units&& units) {
+        auto map_it = _memory_units.find(units.shard);
+        if (map_it == _memory_units.end()) {
+            // Move the first set of semaphore_units to get a copy of
+            // the semaphore pointer and the shard results originates
+            // from.
+            _memory_units.insert_or_assign(units.shard, std::move(units));
```
I'm not following this comment. Can you explain it to me please?

Is this perhaps talking about the fact that we can't just "adopt()" the units of `units` into a default-constructed `semaphore_units` because we want to be able to tie it to the original semaphore? Although looking at the [docs](https://docs.seastar.io/master/classseastar_1_1semaphore__units.html) of `semaphore_units`, it seems like it can't be default-constructed.

ballard26 replied:
`semaphore_units` does have a public default constructor that is being used here. The thing is that it default constructs to a nullptr for the originating semaphore. Hence we're moving the first set of non-empty units into the map.

ballard26 replied:
See https://github.com/scylladb/seastar/blob/master/include/seastar/core/semaphore.hh#L531 . Not sure why it's not in their generated docs.

pgellert commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -37,6 +38,29 @@ using fetch_handler = single_stage_handler<
+struct fetch_memory_units;
+class memory_units_by_ktp {
+public:
+    /// Adds memory units for the provided ktp. Note that any existing memory
+    /// units for the ktp are returned.
+    [[nodiscard]]
+    std::optional<fetch_memory_units>
+    add_units(const model::ktp_with_hash&, fetch_memory_units&&);
```
Should this be called `replace_units` or similar to better reflect the behaviour that this is not "incremental addition" but "replace and return the old units".

ballard26 replied:
Good call, will do

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -639,19 +646,29 @@ static void fill_fetch_responses(
-        resp_it->set(std::move(resp));
+            // TODO: add probe to measure how much of read data is discarded
 
-        if (record_latency) {
-            std::chrono::microseconds fetch_latency
-              = std::chrono::duration_cast<std::chrono::microseconds>(
-                op_context::latency_clock::now() - start_time);
-            octx.rctx.probe().record_fetch_latency(fetch_latency);
```
Why is the fetch latency tracking code moving here? I'm wondering if there are any unintended behaviour changes in the result of the metrics as a result.

ballard26 replied:
Just to make the code a bit cleaner. The metric its self never had a clear region that it measured and shouldn't change much as a result.

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -601,27 +602,33 @@ static void fill_fetch_responses(
         }
 
+        if (!res.has_data()) {
+            // It's possible that data for this partition has already been
+            // written to the response, but is no longer in the partition due to
+            // truncation or otherwise. Hence we move any existing response
+            // units out of the op_context and into the deleter here.
+            mem_units_to_delete.add_units(
+              octx.response_memory_units.remove_units(ktp));
+            resp.records = batch_reader();
+            resp_it->set(std::move(resp));
```
Am I following it correctly that these `remove_units` and `add_units` calls are needed because we may call `fill_fetch_responses` multiple times per `ktp` since with non-polling fetch we try to fetch more and more and re-fill the response.

ballard26 replied:
Yes, we did the same with the polling fetch implementation as well. Each ktp re-fill overrides any previously written result.

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -534,27 +602,33 @@ static void fill_fetch_responses(
             resp.preferred_read_replica = *res.preferred_replica;
         }
 
-        // Aggregate memory_units from all results together to avoid
```
Big +1, this seems like the best way to prevent future bugs as the code evolves.

ballard26 commented on src/v/kafka/server/handlers/fetch.h:
```diff
@@ -66,9 +92,16 @@ struct op_context {
         }
         intrusive_list_hook _hook;
 
+        size_t response_size() {
```
Will add comment

ballard26 commented:
Closing in favor of https://github.com/redpanda-data/redpanda/pull/27828.

ballard26 commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -522,10 +593,7 @@ static void fill_fetch_responses(
          * Cache fetch metadata
```
Good point, will change

ballard26 commented on src/v/kafka/server/tests/fetch_test.cc:
```diff
@@ -1132,3 +1134,79 @@ FIXTURE_TEST(fetch_offset_out_of_range, redpanda_thread_fixture) {
+            }}});
+
+        // create a request
+        kafka::fetch_request_data frq_data;
```
Good point, switching to a designated initializer.
